<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>hdmax2 application • hdmax2</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.4.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="hdmax2 application">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">hdmax2</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">2.0.0.9000</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="active nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><a class="dropdown-item" href="../articles/hdmax2_tutorial.html">hdmax2 application</a></li>
  </ul>
</li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>hdmax2 application</h1>
                        <h4 data-toc-skip class="author">Florence
Pittion, Magali Richard, Olivier Francois</h4>
            
            <h4 data-toc-skip class="date">January, 2024</h4>
      

      <div class="d-none name"><code>hdmax2_tutorial.Rmd</code></div>
    </div>

    
    
<p><strong>THIS VIGNETTE IS CURRENTLY UNDER DEVELOPMENT, SO ITS CONTENT
IS PROVISIONAL.</strong></p>
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p>When a statistical association is observed between an external
exposure
(<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>X</mi><annotation encoding="application/x-tex">X</annotation></semantics></math>)
and an individual outcome
(<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>Y</mi><annotation encoding="application/x-tex">Y</annotation></semantics></math>),
one or more intermediate variables (<strong>M</strong>) (such as gene
expression or epigenetic changes) may mediate this effect. Identifying
and assessing the mediating role of these variables in the effect of
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>X</mi><annotation encoding="application/x-tex">X</annotation></semantics></math>
on
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>Y</mi><annotation encoding="application/x-tex">Y</annotation></semantics></math>
is crucial for deciphering underlying causal mechanisms in
epidemiological and clinical research. This process, known as mediation
analysis, involves studying mediator variables to define the causal
structure between
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>X</mi><annotation encoding="application/x-tex">X</annotation></semantics></math>
and
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>Y</mi><annotation encoding="application/x-tex">Y</annotation></semantics></math>.
The mediated effect, termed the indirect effect, is equal to the effect
of
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>X</mi><annotation encoding="application/x-tex">X</annotation></semantics></math>
on
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>Y</mi><annotation encoding="application/x-tex">Y</annotation></semantics></math>
mediated through (<strong>M</strong>), to distinguish from the direct
effect of
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>X</mi><annotation encoding="application/x-tex">X</annotation></semantics></math>
on
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>Y</mi><annotation encoding="application/x-tex">Y</annotation></semantics></math>
unexplained by (<strong>M</strong>).</p>
<p>The R package <code>hdmax2</code> is dedicated to high-dimensional
mediation analyses. For mediation analyses, the necessary data includes
one (or more) exposure variable(s), a matrix of potential mediators, and
an outcome variable.</p>
<p>The term “high dimensional” refers to the large quantity of potential
mediators among which we seek to identify the actual mediators.</p>
<p>The fundamental concept behind HDMAX2 methods is to use a latent
factor mixed regression model for estimating unobserved latent factors
while conducting high-dimensional association analysis. HDMAX2 also
implements a novel procedure known as the max-squared test to assess the
statistical significance of potential mediators. Finally, HDMAX2 enables
the calculation of an overall indirect effect from a single model that
includes all mediators simultaneously. This approach therefore takes
into account correlations between mediators.</p>
<p>The <code>hdmax2</code> program has been enriched with numerous
features, including the ability to manage various types of variables in
the exposure (continuous, binary, categorical, and multivariate), as
well as the capability to incorporate binary outcomes. This package
enables users to:</p>
<ul>
<li><p>investigate associations between the variables
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>X</mi><annotation encoding="application/x-tex">X</annotation></semantics></math>,
(<strong>M</strong>), and
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>Y</mi><annotation encoding="application/x-tex">Y</annotation></semantics></math>,</p></li>
<li><p>compute the mediated effect for each potential mediator,</p></li>
<li><p>assess the overall indirect effect for the total model,</p></li>
<li><p>and visualize these results.</p></li>
</ul>
<p>The package <code>hdmax2</code> is designed to accept :</p>
<ul>
<li><p>One or several Exposure variables
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>X</mi><annotation encoding="application/x-tex">X</annotation></semantics></math>
consisting of univariate data, which can be continuous, binary or
categorical as well as multivariate exposure. We define
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>n</mi><annotation encoding="application/x-tex">n</annotation></semantics></math>
as the number of samples within the tested cohort. Therefore, each
exposure and outcome variable consists of
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>n</mi><annotation encoding="application/x-tex">n</annotation></semantics></math>
measurements. In the <code>R</code> language, categorical variables are
encoded as factor objects. The function <code><a href="https://rdrr.io/r/base/factor.html" class="external-link">as.factor()</a></code> can be
used to encode categorical variables. The functions
<code><a href="https://rdrr.io/r/base/levels.html" class="external-link">levels()</a></code> and <code><a href="https://rdrr.io/r/base/factor.html" class="external-link">ordered()</a></code> can be used to define
the order of the modalities of categorical variables. By convention,
<code>hdmax2</code> uses the first modality as a reference to calculate
the effects associated with the other modalities of the variable, as
encoded by default in statistical regression models in
<code>R</code>.</p></li>
<li><p>Continuous intermediary variables, denoted as <strong>M</strong>,
are represented as a matrix encompassing potential mediators, such as
methylome or transcriptome molecular features. The matrix
<strong>M</strong> should be entered as a separated input, without
missing values. The intermediary variable matrix <strong>M</strong> is
of dimension
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>n</mi><mo>×</mo><mi>p</mi></mrow><annotation encoding="application/x-tex">n \times p</annotation></semantics></math>,
with
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>p</mi><annotation encoding="application/x-tex">p</annotation></semantics></math>
the total number of intermediate variables.</p></li>
<li><p>The outcome variable, denoted as
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>Y</mi><annotation encoding="application/x-tex">Y</annotation></semantics></math>,
corresponds to a vector of dimension
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>n</mi><annotation encoding="application/x-tex">n</annotation></semantics></math>,
which supports both continuous and binary formats. Continuous and binary
variables must be encoded in numeric format.</p></li>
<li><p>Optional covariates,
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>Z</mi><annotation encoding="application/x-tex">Z</annotation></semantics></math>,
can be included as observed adjustment factors in the model. Please
refer to the helper function section for insight into how to interpret
the additional set of covariates.</p></li>
<li><p><math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>K</mi><annotation encoding="application/x-tex">K</annotation></semantics></math>,
the number of latent factors to be estimated.</p></li>
</ul>
<p>To install the latest version of <code>hdmax2</code>, use the github
repository</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#devtools::install_github("bcm-uga/hdmax2")</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="method-example-multivariate-exposure-and-continuous-outcome">METHOD Example: <em>multivariate exposure</em> and <em>continuous
outcome</em><a class="anchor" aria-label="anchor" href="#method-example-multivariate-exposure-and-continuous-outcome"></a>
</h2>
<p>For this example we use a <strong>two variables exposure</strong>
(continuous and binary), a <strong>continuous outcome</strong> and
<strong>two adjustment factors</strong> (age and gender).</p>
</div>
<div class="section level2">
<h2 id="simulated-dataset">Simulated dataset<a class="anchor" aria-label="anchor" href="#simulated-dataset"></a>
</h2>
<p>We simulated data for 100 samples and 200 potential mediators.</p>
<p>The matrix of potential mediators is an extract of real methylation
data (TCGA PDAC samples).</p>
<p>We define the
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>K</mi><annotation encoding="application/x-tex">K</annotation></semantics></math>
number of estimated latent factors by performing a PCA on potential
mediators matrix. The scree plot criterion looks for the “elbow” in the
curve and selects all components just before the line flattens out,
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>K</mi><annotation encoding="application/x-tex">K</annotation></semantics></math>
is chosen with this procedure.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">simu_data</span> <span class="op">=</span> <span class="fu">hdmax2</span><span class="fu">::</span><span class="va"><a href="../reference/simu_data.html">simu_data</a></span></span></code></pre></div>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Number of Latent factore estimation</span></span>
<span><span class="va">pc</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/prcomp.html" class="external-link">prcomp</a></span><span class="op">(</span><span class="va">simu_data</span><span class="op">$</span><span class="va">M1</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="op">(</span><span class="va">pc</span><span class="op">$</span><span class="va">sdev</span><span class="op">^</span><span class="fl">2</span><span class="op">/</span><span class="fu"><a href="https://rdrr.io/r/base/sum.html" class="external-link">sum</a></span><span class="op">(</span><span class="va">pc</span><span class="op">$</span><span class="va">sdev</span><span class="op">^</span><span class="fl">2</span><span class="op">)</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">10</span><span class="op">]</span>,</span>
<span>     type <span class="op">=</span> <span class="st">"b"</span>,</span>
<span>     xlab <span class="op">=</span> <span class="st">'Principal Component'</span>,</span>
<span>     ylab <span class="op">=</span> <span class="st">"Explained variance"</span><span class="op">)</span></span></code></pre></div>
<p><img src="hdmax2_tutorial_files/figure-html/pca%20for%20latent%20factor-1.png" width="700"></p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="va">K</span><span class="op">=</span><span class="fl">4</span> <span class="co">#pca conclusion : it is better to select too many factors that too few</span></span></code></pre></div>
<div class="section level3">
<h3 id="step-1-run-association-studies">STEP 1: Run association studies<a class="anchor" aria-label="anchor" href="#step-1-run-association-studies"></a>
</h3>
<p>The <code>run_AS</code> function is applied:</p>
<ul>
<li><p>First to estimate latent factors with
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>l</mi><mi>f</mi><mi>m</mi><mi>m</mi></mrow><annotation encoding="application/x-tex">lfmm</annotation></semantics></math>
algorithms.</p></li>
<li><p>Then to identify significant effects of exposure
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>X</mi><annotation encoding="application/x-tex">X</annotation></semantics></math>
on potential mediators in <strong>M</strong> matrix, and significant
effect of potential mediators on outcome
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>Y</mi><annotation encoding="application/x-tex">Y</annotation></semantics></math>.</p></li>
<li><p>And eventually to compute mediation test: <em>max-squared</em>
test.</p></li>
</ul>
<p>The <code>run_AS</code> function takes as inputs:</p>
<ul>
<li>
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>X</mi><annotation encoding="application/x-tex">X</annotation></semantics></math>
for exposure: can be a vector, a factor with
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>n</mi><annotation encoding="application/x-tex">n</annotation></semantics></math>
(samples) elements or a data frame with
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>n</mi><annotation encoding="application/x-tex">n</annotation></semantics></math>
rows and 1 column if univariate or
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>x</mi><annotation encoding="application/x-tex">x</annotation></semantics></math>
columns or different exposure variables if multivariate.</li>
<li>
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>Y</mi><annotation encoding="application/x-tex">Y</annotation></semantics></math>
for outcome: can be a vector or a matrix with
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>n</mi><annotation encoding="application/x-tex">n</annotation></semantics></math>
rows and 1 column.</li>
<li>
<strong>M</strong> for the potential mediators: must be a matrix
with
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>n</mi><annotation encoding="application/x-tex">n</annotation></semantics></math>
rows and
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>p</mi><annotation encoding="application/x-tex">p</annotation></semantics></math>
columns (potential mediators).</li>
<li>
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>K</mi><annotation encoding="application/x-tex">K</annotation></semantics></math>
for the number of latent factors defined earlier: must be integer.</li>
<li>Adjustment factors <em>covar</em> can be included: must be numeric
(data frame or matrix) with
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>n</mi><annotation encoding="application/x-tex">n</annotation></semantics></math>
rows.</li>
</ul>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## multivariate Exposures (continuous + binary)</span></span>
<span><span class="va">X</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span>X1 <span class="op">=</span> <span class="va">simu_data</span><span class="op">$</span><span class="va">X_continuous</span>, X2 <span class="op">=</span> <span class="va">simu_data</span><span class="op">$</span><span class="va">X_binary</span><span class="op">)</span></span>
<span><span class="va">covar</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/cbind.html" class="external-link">cbind</a></span><span class="op">(</span><span class="va">simu_data</span><span class="op">$</span><span class="va">age</span>, <span class="va">simu_data</span><span class="op">$</span><span class="va">gender</span><span class="op">)</span></span>
<span><span class="va">covar</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html" class="external-link">as.data.frame</a></span><span class="op">(</span><span class="va">covar</span><span class="op">)</span></span>
<span></span>
<span><span class="va">hdmax2_step1</span> <span class="op">=</span> <span class="fu">hdmax2</span><span class="fu">::</span><span class="fu"><a href="../reference/run_AS.html">run_AS</a></span><span class="op">(</span>exposure <span class="op">=</span> <span class="va">X</span> ,</span>
<span>                              outcome <span class="op">=</span> <span class="va">simu_data</span><span class="op">$</span><span class="va">Y_continuous</span>,</span>
<span>                              M <span class="op">=</span> <span class="va">simu_data</span><span class="op">$</span><span class="va">M2</span>,</span>
<span>                              K <span class="op">=</span> <span class="va">K</span>,</span>
<span>                              covar <span class="op">=</span> <span class="va">covar</span><span class="op">)</span></span></code></pre></div>
<p>The <code>run_AS</code> function provides an object containing:</p>
<ul>
<li>Results from the two association studies
(<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>P</mi><annotation encoding="application/x-tex">P</annotation></semantics></math>-values,
fscores or zscores),</li>
</ul>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">hdmax2_step1</span><span class="op">$</span><span class="va">AS_1</span><span class="op">$</span><span class="va">pval</span><span class="op">)</span></span>
<span><span class="co">#&gt;    cg00005740    cg00006787    cg00007032    cg00008612    cg00009871 </span></span>
<span><span class="co">#&gt;  3.794621e-88  6.729132e-65 2.716521e-107  9.031578e-75  2.799630e-91 </span></span>
<span><span class="co">#&gt;    cg00009922 </span></span>
<span><span class="co">#&gt;  2.167705e-03</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">hdmax2_step1</span><span class="op">$</span><span class="va">AS_1</span><span class="op">$</span><span class="va">fscores</span><span class="op">)</span></span>
<span><span class="co">#&gt; Response cg00005740.value Response cg00006787.value Response cg00007032.value </span></span>
<span><span class="co">#&gt;               11377.11336                3651.15470               28504.22839 </span></span>
<span><span class="co">#&gt; Response cg00008612.value Response cg00009871.value Response cg00009922.value </span></span>
<span><span class="co">#&gt;                5943.10164               13230.32412                  24.55842</span></span></code></pre></div>
<ul>
<li>Latent factors estimation matrix
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>U</mi><annotation encoding="application/x-tex">U</annotation></semantics></math>
(<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>p</mi><annotation encoding="application/x-tex">p</annotation></semantics></math>
rows and
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>K</mi><annotation encoding="application/x-tex">K</annotation></semantics></math>
columns) from first regression,</li>
</ul>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">hdmax2_step1</span><span class="op">$</span><span class="va">AS_1</span><span class="op">$</span><span class="va">U</span><span class="op">)</span></span>
<span><span class="co">#&gt;            [,1]       [,2]       [,3]        [,4]</span></span>
<span><span class="co">#&gt; [1,] -6.4407431 -1.0637980 -0.6642124 -0.10179492</span></span>
<span><span class="co">#&gt; [2,]  1.1180773  0.7722278  0.3412332  0.03089914</span></span>
<span><span class="co">#&gt; [3,]  3.1673367  1.4633086  0.8123724 -0.07885703</span></span>
<span><span class="co">#&gt; [4,]  0.7052038  0.8200024  0.4004379 -0.07368424</span></span>
<span><span class="co">#&gt; [5,] -4.6385980 -0.7803669 -0.5024476 -0.04347855</span></span>
<span><span class="co">#&gt; [6,]  7.3524954  1.4323450  0.7671163 -0.32313944</span></span></code></pre></div>
<ul>
<li><p><code>run_AS</code> function’s inputs,</p></li>
<li><p>And <em>max-squared</em> test
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>P</mi><annotation encoding="application/x-tex">P</annotation></semantics></math>-values
results.</p></li>
</ul>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">hdmax2_step1</span><span class="op">$</span><span class="va">max2_pvalues</span><span class="op">)</span></span>
<span><span class="co">#&gt;   cg00005740   cg00006787   cg00007032   cg00008612   cg00009871   cg00009922 </span></span>
<span><span class="co">#&gt; 3.930603e-07 1.061875e-02 2.933874e-16 2.099247e-04 1.514348e-12 3.184637e-01</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="selection-of-a-subset-of-mediators">Selection of a subset of mediators<a class="anchor" aria-label="anchor" href="#selection-of-a-subset-of-mediators"></a>
</h3>
<p><em>max-squared</em>
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>P</mi><annotation encoding="application/x-tex">P</annotation></semantics></math>-values
are used in the selection of mediators for the user’s chosen method.</p>
<p>Numerous selection methods are available, such as FDR control. Also
in the context of methylation data, it’s feasible to to identify
aggregated mediator regions (AMR) based on the paired
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>P</mi><annotation encoding="application/x-tex">P</annotation></semantics></math>-values
frome mediation test. Refer to the <em>helper_functions</em> vignette
for assistance FDR control and AMR researching procedure.</p>
<p>For the example, we opted for the top ten most significant
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>P</mi><annotation encoding="application/x-tex">P</annotation></semantics></math>-values
from the <em>max-squared</em> test.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Selecting top 10 mediators</span></span>
<span><span class="va">mediators_top10</span> <span class="op">=</span> <span class="va">simu_data</span><span class="op">$</span><span class="va">M2</span><span class="op">[</span>,<span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/sort.html" class="external-link">sort</a></span><span class="op">(</span><span class="va">hdmax2_step1</span><span class="op">$</span><span class="va">max2_pvalues</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">10</span><span class="op">]</span><span class="op">)</span><span class="op">]</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">mediators_top10</span><span class="op">)</span></span>
<span><span class="co">#&gt;            cg00025981 cg00019093 cg00031759 cg00022633 cg00007032 cg00009871</span></span>
<span><span class="co">#&gt; GSM1051525   2.810690   3.333051   2.681391   2.058144   1.431224   2.355835</span></span>
<span><span class="co">#&gt; GSM1051526   7.263145   7.430116   7.058559   5.781962   5.076067   6.516013</span></span>
<span><span class="co">#&gt; GSM1051527   8.663460   8.633662   8.428823   6.993801   6.293432   7.910915</span></span>
<span><span class="co">#&gt; GSM1051528   7.343681   7.353801   7.121072   5.887332   5.290542   6.760180</span></span>
<span><span class="co">#&gt; GSM1051529   3.567254   4.151531   3.400333   2.631943   1.900969   2.915196</span></span>
<span><span class="co">#&gt; GSM1051530   9.566706  10.000372   9.228616   7.516322   6.406738   8.090836</span></span>
<span><span class="co">#&gt;            cg00028749 cg00014667 cg00005740 cg00012692</span></span>
<span><span class="co">#&gt; GSM1051525   2.491659   2.052389   2.648494   1.032899</span></span>
<span><span class="co">#&gt; GSM1051526   5.744348   5.498459   7.376805   3.708204</span></span>
<span><span class="co">#&gt; GSM1051527   6.707678   6.642283   8.956235   4.604570</span></span>
<span><span class="co">#&gt; GSM1051528   5.714551   5.694672   7.643873   3.889778</span></span>
<span><span class="co">#&gt; GSM1051529   3.106085   2.499390   3.252984   1.383063</span></span>
<span><span class="co">#&gt; GSM1051530   7.654547   6.773123   9.050847   4.678163</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="step-2">STEP 2<a class="anchor" aria-label="anchor" href="#step-2"></a>
</h3>
<p>The function <code>estimate_effect</code> estimate the individual
indirect effect of mediators, but also overall effects of selected
mediators.</p>
<p>The function <code>estimate_effect</code> takes as inputs, step 1
object and selected mediators matrix
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msup><mi>M</mi><mi>S</mi></msup><annotation encoding="application/x-tex">M^S</annotation></semantics></math>
from chosen selection method apply on <em>max-squared</em> test
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>P</mi><annotation encoding="application/x-tex">P</annotation></semantics></math>-values.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">hdmax2_step2</span> <span class="op">=</span> <span class="fu">hdmax2</span><span class="fu">::</span><span class="fu"><a href="../reference/estimate_effect.html">estimate_effect</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">hdmax2_step1</span>,</span>
<span>                                       m <span class="op">=</span> <span class="va">mediators_top10</span><span class="op">)</span></span></code></pre></div>
<p>The function <code>estimate_effect</code> use
<code><a href="https://rdrr.io/pkg/mediation/man/mediate.html" class="external-link">mediation::mediate</a></code> function to obtain several effects
estimation with uncertainty:</p>
<ul>
<li>ACME Average Causal Mediation Effect: corresponding to the indirect
effect</li>
</ul>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># For first exposure variable</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">hdmax2_step2</span><span class="op">$</span><span class="va">effects</span><span class="op">$</span><span class="va">X1</span><span class="op">$</span><span class="va">ACME</span><span class="op">)</span></span>
<span><span class="co">#&gt;           est       CI_2.5     CI_97.5  pval       feat</span></span>
<span><span class="co">#&gt; 1  0.28744276  0.187336837  0.38652463 0.000 cg00025981</span></span>
<span><span class="co">#&gt; 2  0.16583563  0.127190132  0.20654127 0.000 cg00019093</span></span>
<span><span class="co">#&gt; 3  0.15916386  0.062213637  0.25590771 0.002 cg00031759</span></span>
<span><span class="co">#&gt; 4  0.10340946  0.008348204  0.20054958 0.030 cg00022633</span></span>
<span><span class="co">#&gt; 5 -0.18351938 -0.327133190 -0.04509537 0.006 cg00007032</span></span>
<span><span class="co">#&gt; 6  0.02519542 -0.074828002  0.12490948 0.636 cg00009871</span></span>
<span></span>
<span><span class="co"># For first exposure variable</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">hdmax2_step2</span><span class="op">$</span><span class="va">effects</span><span class="op">$</span><span class="va">X2</span><span class="op">$</span><span class="va">ACME</span><span class="op">)</span></span>
<span><span class="co">#&gt;           est      CI_2.5     CI_97.5  pval       feat</span></span>
<span><span class="co">#&gt; 1  0.28270063  0.18067236  0.38560807 0.000 cg00025981</span></span>
<span><span class="co">#&gt; 2  0.16803145  0.12435169  0.21049290 0.000 cg00019093</span></span>
<span><span class="co">#&gt; 3  0.16095023  0.06077006  0.25429286 0.000 cg00031759</span></span>
<span><span class="co">#&gt; 4  0.10135026  0.01052350  0.19546690 0.032 cg00022633</span></span>
<span><span class="co">#&gt; 5 -0.18398511 -0.33110670 -0.02914637 0.020 cg00007032</span></span>
<span><span class="co">#&gt; 6  0.02388381 -0.07319260  0.12097263 0.652 cg00009871</span></span></code></pre></div>
<ul>
<li>PM Proportion Mediate: corresponding to the proportion of the total
effect that is mediated by the mediator</li>
</ul>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># For first exposure variable</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">hdmax2_step2</span><span class="op">$</span><span class="va">effects</span><span class="op">$</span><span class="va">X1</span><span class="op">$</span><span class="va">PM</span><span class="op">)</span></span>
<span><span class="co">#&gt;          est     CI_2.5     CI_97.5  pval       feat</span></span>
<span><span class="co">#&gt; 1 -0.5989691 -0.8054299 -0.39066639 0.000 cg00025981</span></span>
<span><span class="co">#&gt; 2 -0.3457651 -0.4316470 -0.26397633 0.000 cg00019093</span></span>
<span><span class="co">#&gt; 3 -0.3305574 -0.5308632 -0.12927000 0.002 cg00031759</span></span>
<span><span class="co">#&gt; 4 -0.2182789 -0.4155508 -0.01734332 0.030 cg00022633</span></span>
<span><span class="co">#&gt; 5  0.3742992  0.0941027  0.68034763 0.006 cg00007032</span></span>
<span><span class="co">#&gt; 6 -0.0501110 -0.2605593  0.15621632 0.636 cg00009871</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">hdmax2_step2</span><span class="op">$</span><span class="va">effects</span><span class="op">$</span><span class="va">X2</span><span class="op">$</span><span class="va">PM</span><span class="op">)</span></span>
<span><span class="co">#&gt;          est      CI_2.5     CI_97.5  pval       feat</span></span>
<span><span class="co">#&gt; 1 -0.5929589 -0.80198906 -0.37466460 0.000 cg00025981</span></span>
<span><span class="co">#&gt; 2 -0.3518127 -0.43693413 -0.26037291 0.000 cg00019093</span></span>
<span><span class="co">#&gt; 3 -0.3427913 -0.53202931 -0.12643011 0.000 cg00031759</span></span>
<span><span class="co">#&gt; 4 -0.2126285 -0.40759815 -0.02174242 0.032 cg00022633</span></span>
<span><span class="co">#&gt; 5  0.3847555  0.06066514  0.69409402 0.020 cg00007032</span></span>
<span><span class="co">#&gt; 6 -0.0517334 -0.25054803  0.15291725 0.652 cg00009871</span></span></code></pre></div>
<ul>
<li>TE total effect: which is equal to the sum of direct and indirect
effect</li>
</ul>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># For first exposure variable</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">hdmax2_step2</span><span class="op">$</span><span class="va">effects</span><span class="op">$</span><span class="va">X1</span><span class="op">$</span><span class="va">TE</span><span class="op">)</span></span>
<span><span class="co">#&gt;          est     CI_2.5    CI_97.5 pval       feat</span></span>
<span><span class="co">#&gt; 1 -0.4809432 -0.4871167 -0.4746203    0 cg00025981</span></span>
<span><span class="co">#&gt; 2 -0.4805523 -0.4868689 -0.4741050    0 cg00019093</span></span>
<span><span class="co">#&gt; 3 -0.4801520 -0.4864263 -0.4735749    0 cg00031759</span></span>
<span><span class="co">#&gt; 4 -0.4797061 -0.4860125 -0.4731113    0 cg00022633</span></span>
<span><span class="co">#&gt; 5 -0.4793463 -0.4855869 -0.4730238    0 cg00007032</span></span>
<span><span class="co">#&gt; 6 -0.4796200 -0.4859869 -0.4736794    0 cg00009871</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">hdmax2_step2</span><span class="op">$</span><span class="va">effects</span><span class="op">$</span><span class="va">X2</span><span class="op">$</span><span class="va">TE</span><span class="op">)</span></span>
<span><span class="co">#&gt;          est     CI_2.5    CI_97.5 pval       feat</span></span>
<span><span class="co">#&gt; 1 -0.4807088 -0.4867609 -0.4747383    0 cg00025981</span></span>
<span><span class="co">#&gt; 2 -0.4803955 -0.4866038 -0.4742374    0 cg00019093</span></span>
<span><span class="co">#&gt; 3 -0.4798673 -0.4861947 -0.4738398    0 cg00031759</span></span>
<span><span class="co">#&gt; 4 -0.4793114 -0.4851246 -0.4732602    0 cg00022633</span></span>
<span><span class="co">#&gt; 5 -0.4790902 -0.4847672 -0.4730059    0 cg00007032</span></span>
<span><span class="co">#&gt; 6 -0.4792182 -0.4848416 -0.4736264    0 cg00009871</span></span></code></pre></div>
<ul>
<li>ADE Average Direct Effect: which represents the unmediated
effect.</li>
</ul>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># For first exposure variable</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">hdmax2_step2</span><span class="op">$</span><span class="va">effects</span><span class="op">$</span><span class="va">X1</span><span class="op">$</span><span class="va">ADE</span><span class="op">)</span></span>
<span><span class="co">#&gt;          est     CI_2.5    CI_97.5 pval       feat</span></span>
<span><span class="co">#&gt; 1 -0.7683860 -0.8681962 -0.6668683    0 cg00025981</span></span>
<span><span class="co">#&gt; 2 -0.6463879 -0.6860880 -0.6058558    0 cg00019093</span></span>
<span><span class="co">#&gt; 3 -0.6393159 -0.7373529 -0.5430835    0 cg00031759</span></span>
<span><span class="co">#&gt; 4 -0.5831156 -0.6806418 -0.4871739    0 cg00022633</span></span>
<span><span class="co">#&gt; 5 -0.2958269 -0.4341160 -0.1524207    0 cg00007032</span></span>
<span><span class="co">#&gt; 6 -0.5048154 -0.6047025 -0.4038993    0 cg00009871</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">hdmax2_step2</span><span class="op">$</span><span class="va">effects</span><span class="op">$</span><span class="va">X2</span><span class="op">$</span><span class="va">ADE</span><span class="op">)</span></span>
<span><span class="co">#&gt;          est     CI_2.5    CI_97.5 pval       feat</span></span>
<span><span class="co">#&gt; 1 -0.7634094 -0.8643089 -0.6616243    0 cg00025981</span></span>
<span><span class="co">#&gt; 2 -0.6484270 -0.6908482 -0.6031792    0 cg00019093</span></span>
<span><span class="co">#&gt; 3 -0.6408175 -0.7359112 -0.5410028    0 cg00031759</span></span>
<span><span class="co">#&gt; 4 -0.5806617 -0.6733152 -0.4890897    0 cg00022633</span></span>
<span><span class="co">#&gt; 5 -0.2951051 -0.4491536 -0.1459359    0 cg00007032</span></span>
<span><span class="co">#&gt; 6 -0.5031020 -0.6006787 -0.4047876    0 cg00009871</span></span></code></pre></div>
<p>This step also compute Overall effects :</p>
<ul>
<li>OIE (Indirect effect): corresponding to the sum of the indirect
effect associated with all mediators.</li>
</ul>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Estimate OIE for first exposure variable</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">hdmax2_step2</span><span class="op">$</span><span class="va">effects</span><span class="op">$</span><span class="va">X1</span><span class="op">$</span><span class="va">oie</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] -0.027217534  0.096574902  0.104392378  0.005929419 -0.009328689</span></span>
<span><span class="co">#&gt; [6] -0.035804358</span></span>
<span><span class="co"># OIE median for first exposure variable</span></span>
<span><span class="va">hdmax2_step2</span><span class="op">$</span><span class="va">effects</span><span class="op">$</span><span class="va">X1</span><span class="op">$</span><span class="va">oie_med</span></span>
<span><span class="co">#&gt; [1] -0.03725154</span></span>
<span><span class="co"># OIE standard deviation for first exposure variable</span></span>
<span><span class="va">hdmax2_step2</span><span class="op">$</span><span class="va">effects</span><span class="op">$</span><span class="va">X1</span><span class="op">$</span><span class="va">oie_sd</span></span>
<span><span class="co">#&gt; [1] 0.09416536</span></span></code></pre></div>
<ul>
<li>ODE (Direct Effect): corresponding to the effect of exposure
variables on the outcome variable.</li>
</ul>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">hdmax2_step2</span><span class="op">$</span><span class="va">effects</span><span class="op">$</span><span class="va">X1</span><span class="op">$</span><span class="va">ode</span></span>
<span><span class="co">#&gt;      Estimate    Std. Error       t value      Pr(&gt;|t|) </span></span>
<span><span class="co">#&gt; -4.432670e-01  9.192462e-02 -4.822070e+00  6.354078e-06</span></span></code></pre></div>
<ul>
<li>OTE (Total Effect): corresponding to the effect of exposure
variables on the outcome variable when the mediators
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msup><mi>M</mi><mi>S</mi></msup><annotation encoding="application/x-tex">M^S</annotation></semantics></math>
are included in the model.</li>
</ul>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">hdmax2_step2</span><span class="op">$</span><span class="va">effects</span><span class="op">$</span><span class="va">X1</span><span class="op">$</span><span class="va">ote</span></span>
<span><span class="co">#&gt;       Estimate     Std. Error        t value       Pr(&gt;|t|) </span></span>
<span><span class="co">#&gt;  -4.795847e-01   3.166450e-03  -1.514582e+02  4.007925e-113</span></span></code></pre></div>
<p>In addition, function <code>estimate_effect</code> estimates the
intermediary effect sizes
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>a</mi><mi>j</mi></msub><annotation encoding="application/x-tex">a_j</annotation></semantics></math>
and
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>b</mi><mi>j</mi></msub><annotation encoding="application/x-tex">b_j</annotation></semantics></math>
and their standard deviations.</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">hdmax2_step2</span><span class="op">$</span><span class="va">effects</span><span class="op">$</span><span class="va">X1</span><span class="op">$</span><span class="va">xm</span><span class="op">)</span></span>
<span><span class="co">#&gt;   Estimate  Std.Error   t.Value        pValue       feat</span></span>
<span><span class="co">#&gt; 1 2.711291 0.01497632 181.03849 3.517681e-118 cg00025981</span></span>
<span><span class="co">#&gt; 2 1.479219 0.01817362  81.39377  8.455368e-87 cg00019093</span></span>
<span><span class="co">#&gt; 3 2.796769 0.01766869 158.28953 6.864229e-113 cg00031759</span></span>
<span><span class="co">#&gt; 4 2.716945 0.01906228 142.52990 9.221372e-109 cg00022633</span></span>
<span><span class="co">#&gt; 5 3.463424 0.01598202 216.70758 2.852492e-125 cg00007032</span></span>
<span><span class="co">#&gt; 6 3.976139 0.02683199 148.18650 2.711328e-110 cg00009871</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">hdmax2_step2</span><span class="op">$</span><span class="va">effects</span><span class="op">$</span><span class="va">X1</span><span class="op">$</span><span class="va">my</span><span class="op">)</span></span>
<span><span class="co">#&gt;       Estimate  Std.Error    t.Value       pValue       feat</span></span>
<span><span class="co">#&gt; 1  0.104501476 0.01954063  5.3479068 6.453806e-07 cg00025981</span></span>
<span><span class="co">#&gt; 2  0.112511112 0.01484701  7.5780296 2.691362e-11 cg00019093</span></span>
<span><span class="co">#&gt; 3  0.056823617 0.01772318  3.2061759 1.850305e-03 cg00031759</span></span>
<span><span class="co">#&gt; 4  0.037778407 0.01789985  2.1105431 3.752480e-02 cg00022633</span></span>
<span><span class="co">#&gt; 5 -0.052947915 0.02102400 -2.5184506 1.351468e-02 cg00007032</span></span>
<span><span class="co">#&gt; 6  0.006157449 0.01290205  0.4772458 6.343189e-01 cg00009871</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="vizualisation-of-results">Vizualisation of results<a class="anchor" aria-label="anchor" href="#vizualisation-of-results"></a>
</h3>
<p>We propose a set of plots including:</p>
<ul>
<li><p>Mediators ACME Forest plot</p></li>
<li><p>Mediators PM Forest plot</p></li>
<li><p>Comparison of ODE, OIE and OTE</p></li>
<li><p>Mediators effect size representation</p></li>
</ul>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org" class="external-link">ggplot2</a></span><span class="op">)</span></span>
<span><span class="fu">hdmax2</span><span class="fu">::</span><span class="fu"><a href="../reference/plot_hdmax2.html">plot_hdmax2</a></span><span class="op">(</span><span class="va">hdmax2_step2</span>, plot_type<span class="op">=</span> <span class="st">"all_plot"</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "hdmax2 plot for multivariate exposome"</span></span></code></pre></div>
<p><img src="hdmax2_tutorial_files/figure-html/plot%20results%20multivar%20-1.png" width="960"><img src="hdmax2_tutorial_files/figure-html/plot%20results%20multivar%20-2.png" width="960"></p>
<pre><code><span><span class="co">#&gt; $X1</span></span>
<span><span class="co">#&gt; TableGrob (2 x 2) "arrange": 4 grobs</span></span>
<span><span class="co">#&gt;   z     cells    name           grob</span></span>
<span><span class="co">#&gt; 1 1 (1-1,1-1) arrange gtable[layout]</span></span>
<span><span class="co">#&gt; 2 2 (1-1,2-2) arrange gtable[layout]</span></span>
<span><span class="co">#&gt; 3 3 (2-2,1-1) arrange gtable[layout]</span></span>
<span><span class="co">#&gt; 4 4 (2-2,2-2) arrange gtable[layout]</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $X2</span></span>
<span><span class="co">#&gt; TableGrob (2 x 2) "arrange": 4 grobs</span></span>
<span><span class="co">#&gt;   z     cells    name           grob</span></span>
<span><span class="co">#&gt; 1 1 (1-1,1-1) arrange gtable[layout]</span></span>
<span><span class="co">#&gt; 2 2 (1-1,2-2) arrange gtable[layout]</span></span>
<span><span class="co">#&gt; 3 3 (2-2,1-1) arrange gtable[layout]</span></span>
<span><span class="co">#&gt; 4 4 (2-2,2-2) arrange gtable[layout]</span></span></code></pre>
<ul>
<li><p><strong>A</strong> Estimates of indirect effect (ACME) and
<strong>B</strong> proportions of mediated effect (PM) for the top 10
mediators. The effect estimate is represented by a dot and its 95% CI by
the bar. Symbols correspond to the significance cut off of 5% (square
for p-value
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mo>≥</mo><mn>0.05</mn></mrow><annotation encoding="application/x-tex">\geq 0.05</annotation></semantics></math>,
circle p-value
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mo>&lt;</mo><mn>0.05</mn></mrow><annotation encoding="application/x-tex">&lt; 0.05</annotation></semantics></math>).
Colors correspond to the sign of the effect (green for estimated effect
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mo>≤</mo><mn>0</mn></mrow><annotation encoding="application/x-tex">\leq 0</annotation></semantics></math>
, red for estimated effect
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mo>&gt;</mo><mn>0</mn></mrow><annotation encoding="application/x-tex">&gt; 0</annotation></semantics></math>).</p></li>
<li><p><strong>C</strong> Effect sizes of Overall Direct Effect (ODE),
Overall Indirect Effect (OIE) and Overall Total Effect (OTE). Error bars
correspond to standard deviation (ODE and OTE) or confidence interval
(OIE).</p></li>
<li><p><strong>D</strong> Indirect effect sizes for the selected
mediators. Black corresponds to the ACME, violet to the effect of
exposure
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>X</mi><annotation encoding="application/x-tex">X</annotation></semantics></math>
on mediator <strong>M</strong>, and blue corresponds to the effect of
mediator <strong>M</strong> on outcome
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>Y</mi><annotation encoding="application/x-tex">Y</annotation></semantics></math>.</p></li>
</ul>
<p>In the <code>plot_hdmax2</code> function it is possible to produce
the 4-plots set or each individual plot with <em>plot_type</em>
argument.</p>
</div>
</div>
<div class="section level2">
<h2 id="helper-functions">Helper Functions<a class="anchor" aria-label="anchor" href="#helper-functions"></a>
</h2>
<p>In this vignette, we also provide a series of helper function to
process the data.</p>
<div class="section level3">
<h3 id="how-to-analyse-agregated-methylated-regions">How to analyse agregated methylated regions ?<a class="anchor" aria-label="anchor" href="#how-to-analyse-agregated-methylated-regions"></a>
</h3>
<p>We simulated 100 samples and 200 potential DNA methylation mediators,
with various a binary exposure (smoking status of mothers) and
continuous outcomes (birth weight).</p>
<div class="section level4">
<h4 id="identifying-aggregated-mediator-regions-amr">Identifying aggregated mediator regions (AMR)<a class="anchor" aria-label="anchor" href="#identifying-aggregated-mediator-regions-amr"></a>
</h4>
<p>Identify Aggregated Methylated Regions (AMR) with
<code>AMR_search</code> function which uses from the
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>P</mi><annotation encoding="application/x-tex">P</annotation></semantics></math>-values
from <em>max-squared</em> test compute in <code>run_AS</code> function
and using a adapted <em>comb-p</em> method (<em>comb-p</em> is a tool
that manipulates BED files of possibly irregularly spaced
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>P</mi><annotation encoding="application/x-tex">P</annotation></semantics></math>-values
and calculates auto-correlation, combines adjacent
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>P</mi><annotation encoding="application/x-tex">P</annotation></semantics></math>-values,
performs false discovery adjustment, finds regions of enrichment and
assigns significance to those regions). AMR identification could be
useful in Epigenomic Wide Association Studies (EWAS) when single CpG
mediation is unsuccessful, or could be complementary analysis.</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">helper_ex</span> <span class="op">=</span> <span class="fu">hdmax2</span><span class="fu">::</span><span class="va"><a href="../reference/helper_ex.html">helper_ex</a></span></span>
<span></span>
<span><span class="va">pc</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/prcomp.html" class="external-link">prcomp</a></span><span class="op">(</span><span class="va">helper_ex</span><span class="op">$</span><span class="va">methylation</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="op">(</span><span class="va">pc</span><span class="op">$</span><span class="va">sdev</span><span class="op">^</span><span class="fl">2</span><span class="op">/</span><span class="fu"><a href="https://rdrr.io/r/base/sum.html" class="external-link">sum</a></span><span class="op">(</span><span class="va">pc</span><span class="op">$</span><span class="va">sdev</span><span class="op">^</span><span class="fl">2</span><span class="op">)</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">15</span><span class="op">]</span>,</span>
<span>     xlab <span class="op">=</span> <span class="st">'Principal Component'</span>,</span>
<span>     ylab <span class="op">=</span> <span class="st">"Explained variance"</span>,</span>
<span>     col <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">3</span><span class="op">)</span>, <span class="fl">2</span>, <span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">16</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="hdmax2_tutorial_files/figure-html/AMR%20-1.png" width="700"></p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># chosen number of dimension</span></span>
<span><span class="va">K</span><span class="op">=</span><span class="fl">5</span></span>
<span></span>
<span><span class="co">## run hdmax2 step1</span></span>
<span><span class="va">hdmax2_step1</span> <span class="op">=</span> <span class="fu">hdmax2</span><span class="fu">::</span><span class="fu"><a href="../reference/run_AS.html">run_AS</a></span><span class="op">(</span>exposure <span class="op">=</span> <span class="va">helper_ex</span><span class="op">$</span><span class="va">exposure</span>,</span>
<span>                              outcome <span class="op">=</span> <span class="va">helper_ex</span><span class="op">$</span><span class="va">phenotype</span>,</span>
<span>                              M <span class="op">=</span> <span class="va">helper_ex</span><span class="op">$</span><span class="va">methylation</span>, </span>
<span>                              K <span class="op">=</span> <span class="va">K</span><span class="op">)</span></span>
<span></span>
<span><span class="co">##Detecting AMR</span></span>
<span><span class="va">seed</span> <span class="op">=</span> <span class="fl">0.7</span> <span class="co">#Careful to change this parameter when working with real data</span></span>
<span><span class="va">res.amr_search</span> <span class="op">=</span> <span class="fu">hdmax2</span><span class="fu">::</span><span class="fu"><a href="../reference/AMR_search.html">AMR_search</a></span><span class="op">(</span>chr <span class="op">=</span> <span class="va">helper_ex</span><span class="op">$</span><span class="va">annotation</span><span class="op">$</span><span class="va">chr</span>,</span>
<span>                                    start <span class="op">=</span> <span class="va">helper_ex</span><span class="op">$</span><span class="va">annotation</span><span class="op">$</span><span class="va">start</span>,</span>
<span>                                    end <span class="op">=</span> <span class="va">helper_ex</span><span class="op">$</span><span class="va">annotation</span><span class="op">$</span><span class="va">end</span>,</span>
<span>                                    pval <span class="op">=</span> <span class="va">hdmax2_step1</span><span class="op">$</span><span class="va">max2_pvalues</span>,</span>
<span>                                    cpg <span class="op">=</span> <span class="va">helper_ex</span><span class="op">$</span><span class="va">annotation</span><span class="op">$</span><span class="va">cpg</span>,</span>
<span>                                    seed <span class="op">=</span> <span class="va">seed</span>, </span>
<span>                                    nCores <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span>
<span></span>
<span><span class="va">res.amr_search</span><span class="op">$</span><span class="va">res</span></span>
<span><span class="co">#&gt;   chr  start    end          p        fdr</span></span>
<span><span class="co">#&gt; 3   1 864893 865027 0.03604853 0.05184658</span></span>
<span><span class="co">#&gt; 2   1 858380 858381 0.03901609 0.05184658</span></span>
<span><span class="co">#&gt; 4   1 886431 886553 0.04077365 0.05184658</span></span>
<span><span class="co">#&gt; 5   1 896773 896774 0.05000351 0.05184658</span></span>
<span><span class="co">#&gt; 1   1 854778 855056 0.05184658 0.05184658</span></span>
<span></span>
<span><span class="va">res.arm_build</span> <span class="op">=</span> <span class="fu">hdmax2</span><span class="fu">::</span><span class="fu"><a href="../reference/AMR_build.html">AMR_build</a></span><span class="op">(</span><span class="va">res.amr_search</span>, </span>
<span>                                  methylation <span class="op">=</span> <span class="va">helper_ex</span><span class="op">$</span><span class="va">methylation</span>, </span>
<span>                                  nb_cpg <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span>
<span></span>
<span><span class="co">#List of DMR selected</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">res.arm_build</span><span class="op">$</span><span class="va">res</span><span class="op">)</span></span>
<span><span class="co">#&gt;    DMR chr  start    end          p        fdr nb</span></span>
<span><span class="co">#&gt; 1 DMR1   1 886431 886553 0.04077365 0.05184658  3</span></span>
<span><span class="co">#&gt; 2 DMR2   1 854778 855056 0.05184658 0.05184658  6</span></span>
<span></span>
<span><span class="co">##CpG in the DMR</span></span>
<span><span class="va">res.arm_build</span><span class="op">$</span><span class="va">CpG_for_each_AMR</span></span>
<span><span class="co">#&gt; $DMR1</span></span>
<span><span class="co">#&gt; [1] "cg20322444" "cg07983659" "cg11733071"</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $DMR2</span></span>
<span><span class="co">#&gt; [1] "cg08029603" "cg16347928" "cg22699361" "cg03890988" "cg21786289"</span></span>
<span><span class="co">#&gt; [6] "cg13897241"</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="quantifying-indirect-effects">Quantifying indirect effects<a class="anchor" aria-label="anchor" href="#quantifying-indirect-effects"></a>
</h4>
<p>Like with single mediators analysis <em>esimate_effect</em> function
could be use to estimate different effects of AMR. Also
<code>plot_hdmax2</code> can be applied to step2 results.</p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## run hdmax2 step2</span></span>
<span><span class="va">object</span> <span class="op">=</span> <span class="va">hdmax2_step1</span></span>
<span><span class="va">m</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">as.matrix</a></span><span class="op">(</span><span class="va">res.arm_build</span><span class="op">$</span><span class="va">AMR_mean</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## selected mediators effects estimation</span></span>
<span></span>
<span><span class="va">hdmax2_step2</span> <span class="op">=</span> <span class="fu">hdmax2</span><span class="fu">::</span><span class="fu"><a href="../reference/estimate_effect.html">estimate_effect</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">hdmax2_step1</span>,</span>
<span>                                       m <span class="op">=</span> <span class="va">m</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org" class="external-link">ggplot2</a></span><span class="op">)</span></span>
<span><span class="fu">hdmax2</span><span class="fu">::</span><span class="fu"><a href="../reference/plot_hdmax2.html">plot_hdmax2</a></span><span class="op">(</span><span class="va">hdmax2_step2</span>, plot_type<span class="op">=</span> <span class="st">"all_plot"</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "hdmax2 plot for univariate exposome"</span></span></code></pre></div>
<p><img src="hdmax2_tutorial_files/figure-html/step2-1.png" width="960"></p>
<pre><code><span><span class="co">#&gt; TableGrob (2 x 2) "arrange": 4 grobs</span></span>
<span><span class="co">#&gt;   z     cells    name           grob</span></span>
<span><span class="co">#&gt; 1 1 (1-1,1-1) arrange gtable[layout]</span></span>
<span><span class="co">#&gt; 2 2 (1-1,2-2) arrange gtable[layout]</span></span>
<span><span class="co">#&gt; 3 3 (2-2,1-1) arrange gtable[layout]</span></span>
<span><span class="co">#&gt; 4 4 (2-2,2-2) arrange gtable[layout]</span></span></code></pre>
</div>
</div>
<div class="section level3">
<h3 id="how-to-select-potential-mediators-using-q-values-and-fdr-control">How to select potential mediators using q-values and FDR control
?<a class="anchor" aria-label="anchor" href="#how-to-select-potential-mediators-using-q-values-and-fdr-control"></a>
</h3>
<p>Several methods can be use to select potential mediators from step1
results of our method, in our main use cases we simply select top 10
mediators to simplify the narrative. Among available methods to select
mediators from mediation test
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>P</mi><annotation encoding="application/x-tex">P</annotation></semantics></math>-values,
we can use FDR (False Discovery Rate) control.
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>Q</mi><annotation encoding="application/x-tex">Q</annotation></semantics></math>-value
is obtained from <em>max-squared</em> test
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>P</mi><annotation encoding="application/x-tex">P</annotation></semantics></math>-values
from step 1 with <code><a href="../reference/hdmax2_qvalue.html">hdmax2::hdmax2_qvalue</a></code> function then
bounded by chosen threshold.</p>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## run hdmax2 step1</span></span>
<span><span class="va">hdmax2_step1</span> <span class="op">=</span> <span class="fu">hdmax2</span><span class="fu">::</span><span class="fu"><a href="../reference/run_AS.html">run_AS</a></span><span class="op">(</span>exposure <span class="op">=</span> <span class="va">simu_data</span><span class="op">$</span><span class="va">X_binary</span>,</span>
<span>                 outcome <span class="op">=</span> <span class="va">simu_data</span><span class="op">$</span><span class="va">Y_continuous</span>,</span>
<span>                 M <span class="op">=</span> <span class="va">simu_data</span><span class="op">$</span><span class="va">M1</span>, </span>
<span>                 K <span class="op">=</span> <span class="va">K</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## Select candidate mediator  </span></span>
<span><span class="va">q_values</span> <span class="op">=</span> <span class="fu">hdmax2</span><span class="fu">::</span><span class="fu"><a href="../reference/hdmax2_qvalue.html">hdmax2_qvalue</a></span><span class="op">(</span><span class="va">hdmax2_step1</span><span class="op">$</span><span class="va">max2_pvalues</span><span class="op">)</span></span>
<span><span class="va">candidate_mediator</span> <span class="op">&lt;-</span> <span class="va">q_values</span><span class="op">$</span><span class="va">qv</span><span class="op">[</span><span class="va">q_values</span><span class="op">$</span><span class="va">qv</span><span class="op">&lt;=</span> <span class="fl">0.2</span><span class="op">]</span> <span class="co"># In this example, we will consider FDR levels &lt;20%.</span></span>
<span><span class="va">candidate_mediator</span></span>
<span><span class="co">#&gt;   cg00019093   cg00025981 </span></span>
<span><span class="co">#&gt; 2.175678e-06 1.143605e-02</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="how-to-transform-a-categorial-variable-into-an-ordered-factor">How to transform a categorial variable into an ordered factor?<a class="anchor" aria-label="anchor" href="#how-to-transform-a-categorial-variable-into-an-ordered-factor"></a>
</h3>
<p>When categorical variables are used as exposure variables,
<code>hdmax2</code> uses the first category as a reference (intercept)
to calculate the effects associated with the variable’s other
categories. The functions <code>HDMAX2::run_AS</code> and
<code>HDMAX2::estimate_effect</code> will transform the character vector
you have used (if any) into a factor, with an arbitrary ordering of the
categories. If order is important to you, here’s a simple way to turn
your character vector into a factor and order the categories as you
wish. You can then use this variable as input to the <code>hdmax2</code>
functions.</p>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># inverse alphabetical order for example</span></span>
<span><span class="va">X</span> <span class="op">=</span> <span class="va">simu_data</span><span class="op">$</span><span class="va">X_categorial</span></span>
<span><span class="va">X</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">factor</a></span><span class="op">(</span><span class="va">X</span>, ordered <span class="op">=</span> <span class="cn">TRUE</span>, levels <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"D"</span>,<span class="st">"C"</span>,<span class="st">"B"</span>,<span class="st">"A"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="how-to-handle-adding-an-additional-set-of-covariates-to-the-second-association-study">How to handle adding an additional set of covariates to the second
association study?<a class="anchor" aria-label="anchor" href="#how-to-handle-adding-an-additional-set-of-covariates-to-the-second-association-study"></a>
</h3>
<p>It is possible to add a second adjustment factors set in association
study between potential mediators and outcome if it makes sense from a
biological standpoint. Nevertheless, this additional set of adjustment
factors for the second association study must include the adjustment
factors from the first association study as design in
<code>run_AS</code> function.</p>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># First adjustment factors set</span></span>
<span><span class="va">covar</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html" class="external-link">as.data.frame</a></span><span class="op">(</span><span class="va">helper_ex</span><span class="op">$</span><span class="va">covariables</span><span class="op">[</span>,<span class="fl">1</span><span class="op">:</span><span class="fl">2</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="co"># Second adjustment factors set</span></span>
<span><span class="va">suppl_covar</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html" class="external-link">as.data.frame</a></span><span class="op">(</span><span class="va">helper_ex</span><span class="op">$</span><span class="va">covariables</span><span class="op">[</span>,<span class="fl">3</span><span class="op">:</span><span class="fl">4</span><span class="op">]</span><span class="op">)</span></span>
<span></span>
<span><span class="va">hdmax2_step1</span> <span class="op">=</span> <span class="fu">hdmax2</span><span class="fu">::</span><span class="fu"><a href="../reference/run_AS.html">run_AS</a></span><span class="op">(</span>exposure <span class="op">=</span> <span class="va">helper_ex</span><span class="op">$</span><span class="va">exposure</span>,</span>
<span>                              outcome <span class="op">=</span> <span class="va">helper_ex</span><span class="op">$</span><span class="va">phenotype</span>,</span>
<span>                              M <span class="op">=</span> <span class="va">helper_ex</span><span class="op">$</span><span class="va">methylation</span>,</span>
<span>                              K <span class="op">=</span> <span class="va">K</span>,</span>
<span>                              covar <span class="op">=</span> <span class="va">covar</span>,</span>
<span>                              suppl_covar <span class="op">=</span> <span class="va">suppl_covar</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
</div>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Florence Pittion, Magali Richard, Olivier Francois, Basile Jumentier.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.0.</p>
</div>

    </footer>
</div>





  </body>
</html>
