<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="hdmax2">
<title>hdmax2 application • hdmax2</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.11/clipboard.min.js" integrity="sha512-7O5pXpc0oCRrxk8RUfDYFgn0nO1t+jLuIOQdOMRp4APB7uZ4vSjspzp5y6YDtDs4VzUSTbWzBFZ/LKJhnyFOKw==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="hdmax2 application">
<meta property="og:description" content="hdmax2">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light" data-bs-theme="light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">hdmax2</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">2.0.0.9000</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="active nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/hdmax2_tutorial.html">hdmax2 application</a>
  </div>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav"></ul>
</div>

    
  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>hdmax2 application</h1>
                        <h4 data-toc-skip class="author">Florence
Pittion, Magali Richard, Olivier Francois</h4>
            
            <h4 data-toc-skip class="date">January, 2024</h4>
      
      
      <div class="d-none name"><code>hdmax2_tutorial.Rmd</code></div>
    </div>

    
    
<p><strong>THIS VIGNETTE IS CURRENTLY UNDER DEVELOPMENT, SO ITS CONTENT
IS PROVISIONAL.</strong></p>
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p>When a statistical association is observed between an external
exposure (<span class="math inline">\(X\)</span>) and an individual
outcome (<span class="math inline">\(Y\)</span>), one or more
intermediate variables (<strong>M</strong>) (such as gene expression or
epigenetic changes) may mediate this effect. Identifying and assessing
the mediating role of these variables in the effect of <span class="math inline">\(X\)</span> on <span class="math inline">\(Y\)</span> is crucial for deciphering underlying
causal mechanisms in epidemiological and clinical research. This
process, known as mediation analysis, involves studying mediator
variables to define the causal structure between <span class="math inline">\(X\)</span> and <span class="math inline">\(Y\)</span>. The mediated effect, termed the
indirect effect, is equal to the effect of <span class="math inline">\(X\)</span> on <span class="math inline">\(Y\)</span> mediated through (<strong>M</strong>),
to distinguish from the direct effect of <span class="math inline">\(X\)</span> on <span class="math inline">\(Y\)</span> unexplained by
(<strong>M</strong>).</p>
<p>The R package <code>hdmax2</code> is dedicated to high-dimensional
mediation analyses. For mediation analyses, the necessary data includes
one (or more) exposure variable(s), a matrix of potential mediators, and
an outcome variable.</p>
<p>The term “high dimensional” refers to the large quantity of potential
mediators among which we seek to identify the actual mediators.</p>
<p>The fundamental concept behind HDMAX2 methods is to use a latent
factor mixed regression model for estimating unobserved latent factors
while conducting high-dimensional association analysis. HDMAX2 also
implements a novel procedure known as the max-squared test to assess the
statistical significance of potential mediators. Finally, HDMAX2 enables
the calculation of an overall indirect effect from a single model that
includes all mediators simultaneously. This approach therefore takes
into account correlations between mediators.</p>
<p>The <code>hdmax2</code> program has been enriched with numerous
features, including the ability to manage various types of variables in
the exposure (continuous, binary, categorical, and multivariate), as
well as the capability to incorporate binary outcomes. This package
enables users to:</p>
<ul>
<li><p>investigate associations between the variables <span class="math inline">\(X\)</span>, (<strong>M</strong>), and <span class="math inline">\(Y\)</span>,</p></li>
<li><p>compute the mediated effect for each potential mediator,</p></li>
<li><p>assess the overall indirect effect for the total model,</p></li>
<li><p>and visualize these results.</p></li>
</ul>
<p>The package <code>hdmax2</code> is designed to accept :</p>
<ul>
<li><p>One or several Exposure variables <span class="math inline">\(X\)</span> consisting of univariate data, which
can be continuous, binary or categorical as well as multivariate
exposure. We define <span class="math inline">\(n\)</span> as the number
of samples within the tested cohort. Therefore, each exposure and
outcome variable consists of <span class="math inline">\(n\)</span>
measurements. In the <code>R</code> language, categorical variables are
encoded as factor objects. The function <code><a href="https://rdrr.io/r/base/factor.html" class="external-link">as.factor()</a></code> can be
used to encode categorical variables. The functions
<code><a href="https://rdrr.io/r/base/levels.html" class="external-link">levels()</a></code> and <code><a href="https://rdrr.io/r/base/factor.html" class="external-link">ordered()</a></code> can be used to define
the order of the modalities of categorical variables. By convention,
<code>hdmax2</code> uses the first modality as a reference to calculate
the effects associated with the other modalities of the variable, as
encoded by default in statistical regression models in
<code>R</code>.</p></li>
<li><p>Continuous intermediary variables, denoted as <strong>M</strong>,
are represented as a matrix encompassing potential mediators, such as
methylome or transcriptome molecular features. The matrix
<strong>M</strong> should be entered as a separated input, without
missing values. The intermediary variable matrix <strong>M</strong> is
of dimension <span class="math inline">\(n \times p\)</span>, with <span class="math inline">\(p\)</span> the total number of intermediate
variables.</p></li>
<li><p>The outcome variable, denoted as <span class="math inline">\(Y\)</span>, corresponds to a vector of dimension
<span class="math inline">\(n\)</span>, which supports both continuous
and binary formats. Continuous and binary variables must be encoded in
numeric format.</p></li>
<li><p>Optional covariates, <span class="math inline">\(Z\)</span>, can
be included as observed adjustment factors in the model. Please refer to
the helper function section for insight into how to interpret the
additional set of covariates.</p></li>
<li><p><span class="math inline">\(K\)</span>, the number of latent
factors to be estimated.</p></li>
</ul>
<p>To install the latest version of <code>hdmax2</code>, use the github
repository</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#devtools::install_github("bcm-uga/hdmax2")</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="method">METHOD<a class="anchor" aria-label="anchor" href="#method"></a>
</h2>
<div class="section level3">
<h3 id="step-1-run-association-studies">STEP 1: Run association studies<a class="anchor" aria-label="anchor" href="#step-1-run-association-studies"></a>
</h3>
<p>The <code>run_AS</code> function is applied:</p>
<ul>
<li><p>First to estimate latent factors with <span class="math inline">\(lfmm\)</span> algorithms.</p></li>
<li><p>Then to identify significant effects of exposure <span class="math inline">\(X\)</span> on potential mediators in
<strong>M</strong> matrix, and significant effect of potential mediators
on outcome <span class="math inline">\(Y\)</span>.</p></li>
<li><p>And eventually to compute mediation test: <em>max-squared</em>
test.</p></li>
</ul>
<p>The <code>run_AS</code> function takes as inputs:</p>
<ul>
<li>
<span class="math inline">\(X\)</span> for exposure: can be a
vector, a factor with <span class="math inline">\(n\)</span> (samples)
elements or a data frame with <span class="math inline">\(n\)</span>
rows and 1 column if univariate or <span class="math inline">\(x\)</span> columns or different exposure variables
if multivariate.</li>
<li>
<span class="math inline">\(Y\)</span> for outcome: can be a vector
or a matrix with <span class="math inline">\(n\)</span> rows and 1
column.</li>
<li>
<strong>M</strong> for the potential mediators: must be a matrix
with <span class="math inline">\(n\)</span> rows and <span class="math inline">\(p\)</span> columns (potential mediators).</li>
<li>
<span class="math inline">\(K\)</span> for the number of latent
factors defined earlier: must be integer.</li>
<li>Adjustment factors <em>covar</em> can be included: must be numeric
(data frame or matrix) with <span class="math inline">\(n\)</span>
rows.</li>
</ul>
<p>The <code>run_AS</code> function provides an object containing:</p>
<ul>
<li><p>Results from the two association studies (<span class="math inline">\(P\)</span>-values, fscores, zscores),</p></li>
<li><p>Latent factors estimation matrix <span class="math inline">\(U\)</span> (<span class="math inline">\(p\)</span>
rows and <span class="math inline">\(K\)</span> columns) from first
regression,</p></li>
<li><p><code>run_AS</code> function’s inputs,</p></li>
<li><p>And <em>max-squared</em> test <span class="math inline">\(P\)</span>-values results.</p></li>
</ul>
</div>
<div class="section level3">
<h3 id="selection-of-a-subset-of-mediators">Selection of a subset of mediators<a class="anchor" aria-label="anchor" href="#selection-of-a-subset-of-mediators"></a>
</h3>
<p><em>max-squared</em> <span class="math inline">\(P\)</span>-values
are used in the selection of mediators for the user’s chosen method.</p>
<p>Numerous selection methods are available, such as FDR control. Also
in the context of methylation data, it’s feasible to to identify
aggregated mediator regions (AMR) based on the paired <span class="math inline">\(P\)</span>-values frome mediation test. Refer to
the <em>helper_functions</em> vignette for assistance FDR control and
AMR researching procedure.</p>
<p>For the two examples below, we opted for the top ten most significant
<span class="math inline">\(P\)</span>-values from the
<em>max-squared</em> test as an example.</p>
</div>
<div class="section level3">
<h3 id="step-2">STEP 2<a class="anchor" aria-label="anchor" href="#step-2"></a>
</h3>
<p>The function <code>estimate_effect</code> estimate the individual
indirect effect of mediators, but also overall effects of selected
mediators.</p>
<p>The function <code>estimate_effect</code> takes as inputs, step 1
object and selected mediators matrix <span class="math inline">\(M^S\)</span> from chosen selection method apply on
<em>max-squared</em> test <span class="math inline">\(P\)</span>-values.</p>
<p>The function <code>estimate_effect</code> use
<code><a href="https://rdrr.io/pkg/mediation/man/mediate.html" class="external-link">mediation::mediate</a></code> function to obtain several effects
estimation with uncertainty:</p>
<ul>
<li><p>ACME Average Causal Mediation Effect: corresponding to the
indirect effect</p></li>
<li><p>PM Proportion Mediate: corresponding to the proportion of the
total effect that is mediated by the mediator</p></li>
<li><p>TE total effect: which is equal to the sum of direct and indirect
effect</p></li>
<li><p>ADE Average Direct Effect: which represents the unmediated
effect.</p></li>
</ul>
<p>This step also compute Overall effects :</p>
<ul>
<li><p>OIE (Indirect effect): corresponding to the sum of the indirect
effect associated with all mediators.</p></li>
<li><p>ODE (Direct Effect): corresponding to the effect of exposure
variables on the outcome variable.</p></li>
<li><p>OTE (Total Effect): corresponding to the effect of exposure
variables on the outcome variable when the mediators <span class="math inline">\(M^S\)</span> are included in the model.</p></li>
</ul>
<p>In addition, function <code>estimate_effect</code> estimates the
intermediary effect sizes <span class="math inline">\(a_j\)</span> and
<span class="math inline">\(b_j\)</span> and their standard
deviations.</p>
</div>
<div class="section level3">
<h3 id="vizualisation-of-results">Vizualisation of results<a class="anchor" aria-label="anchor" href="#vizualisation-of-results"></a>
</h3>
<p>We propose a set of plots including:</p>
<ul>
<li><p>Mediators ACME Forest plot</p></li>
<li><p>Mediators PM Forest plot</p></li>
<li><p>Comparison of ODE, OIE and OTE</p></li>
<li><p>Mediators effect size representation</p></li>
</ul>
</div>
</div>
<div class="section level2">
<h2 id="simulated-dataset">Simulated dataset<a class="anchor" aria-label="anchor" href="#simulated-dataset"></a>
</h2>
<p>We simulated data for 100 samples and 500 potential mediators.</p>
<p>The matrix of potential mediators is an extract of real methylation
data (TCGA PDAC samples).</p>
<p>We define the <span class="math inline">\(K\)</span> number of
estimated latent factors by performing a PCA on potential mediators
matrix. The scree plot criterion looks for the “elbow” in the curve and
selects all components just before the line flattens out, <span class="math inline">\(K\)</span> is chosen with this procedure.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">simu_data</span> <span class="op">=</span> <span class="fu">hdmax2</span><span class="fu">::</span><span class="va"><a href="../reference/simu_data.html">simu_data</a></span></span></code></pre></div>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Number of Latent factore estimation</span></span>
<span><span class="va">pc</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/prcomp.html" class="external-link">prcomp</a></span><span class="op">(</span><span class="va">simu_data</span><span class="op">$</span><span class="va">M1</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="op">(</span><span class="va">pc</span><span class="op">$</span><span class="va">sdev</span><span class="op">^</span><span class="fl">2</span><span class="op">/</span><span class="fu"><a href="https://rdrr.io/r/base/sum.html" class="external-link">sum</a></span><span class="op">(</span><span class="va">pc</span><span class="op">$</span><span class="va">sdev</span><span class="op">^</span><span class="fl">2</span><span class="op">)</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">10</span><span class="op">]</span>,</span>
<span>     type <span class="op">=</span> <span class="st">"b"</span>,</span>
<span>     xlab <span class="op">=</span> <span class="st">'Principal Component'</span>,</span>
<span>     ylab <span class="op">=</span> <span class="st">"Explained variance"</span><span class="op">)</span></span></code></pre></div>
<p><img src="hdmax2_tutorial_files/figure-html/pca%20for%20latent%20factor-1.png" width="700"></p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="va">K</span><span class="op">=</span><span class="fl">4</span> <span class="co">#pca conclusion : it is better to select too many factors that too few</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="first-example-binary-exposure-and-continuous-outcome">First Example: <em>Binary exposure</em> and <em>continuous
outcome</em><a class="anchor" aria-label="anchor" href="#first-example-binary-exposure-and-continuous-outcome"></a>
</h2>
<p>For this example we use a <strong>binary exposure</strong>, a
<strong>continuous outcome</strong> and <strong>two adjustment
factors</strong> (age and gender).</p>
<div class="section level3">
<h3 id="step-1">Step 1<a class="anchor" aria-label="anchor" href="#step-1"></a>
</h3>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">covar</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/cbind.html" class="external-link">cbind</a></span><span class="op">(</span><span class="va">simu_data</span><span class="op">$</span><span class="va">age</span>, <span class="va">simu_data</span><span class="op">$</span><span class="va">gender</span><span class="op">)</span></span>
<span><span class="va">covar</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html" class="external-link">as.data.frame</a></span><span class="op">(</span><span class="va">covar</span><span class="op">)</span></span>
<span></span>
<span><span class="va">hdmax2_step1</span> <span class="op">=</span> <span class="fu">hdmax2</span><span class="fu">::</span><span class="fu"><a href="../reference/run_AS.html">run_AS</a></span><span class="op">(</span>exposure <span class="op">=</span> <span class="va">simu_data</span><span class="op">$</span><span class="va">X_binary</span>,</span>
<span>                              outcome  <span class="op">=</span> <span class="va">simu_data</span><span class="op">$</span><span class="va">Y_continuous</span>,</span>
<span>                              M <span class="op">=</span> <span class="va">simu_data</span><span class="op">$</span><span class="va">M1</span>, </span>
<span>                              K <span class="op">=</span> <span class="va">K</span>,</span>
<span>                              covar <span class="op">=</span> <span class="va">covar</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="selection-of-a-subset-of-mediators-1">Selection of a subset of mediators<a class="anchor" aria-label="anchor" href="#selection-of-a-subset-of-mediators-1"></a>
</h3>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Selecting top 10 mediators</span></span>
<span><span class="va">mediators_top10</span> <span class="op">=</span> <span class="va">simu_data</span><span class="op">$</span><span class="va">M1</span><span class="op">[</span>,<span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/sort.html" class="external-link">sort</a></span><span class="op">(</span><span class="va">hdmax2_step1</span><span class="op">$</span><span class="va">max2_pvalues</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">10</span><span class="op">]</span><span class="op">)</span><span class="op">]</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">mediators_top10</span><span class="op">)</span></span>
<span><span class="co">#&gt;            cg00022633 cg00019093 cg00025981 cg00031759 cg00035636 cg00049102</span></span>
<span><span class="co">#&gt; GSM1051525   1.480526   1.091212  0.8611761   1.096765   1.498758   2.029784</span></span>
<span><span class="co">#&gt; GSM1051526   5.463677   3.474162  3.7370006   3.996897   1.481787   2.000932</span></span>
<span><span class="co">#&gt; GSM1051527   5.334700   3.369454  3.6640206   3.889718   1.439573   1.891203</span></span>
<span><span class="co">#&gt; GSM1051528   5.225207   3.289293  3.6175162   3.808743   1.443054   1.816579</span></span>
<span><span class="co">#&gt; GSM1051529   1.691596   1.260262  0.9876545   1.249065   1.603649   2.217188</span></span>
<span><span class="co">#&gt; GSM1051530   6.307245   4.088710  4.2104976   4.571531   2.310486   2.725880</span></span>
<span><span class="co">#&gt;            cg00028749 cg00049616 cg00026603 cg00703120</span></span>
<span><span class="co">#&gt; GSM1051525  0.7984107   1.767392   1.840380  0.8521711</span></span>
<span><span class="co">#&gt; GSM1051526  3.9441627   1.743678   3.914640  0.8315540</span></span>
<span><span class="co">#&gt; GSM1051527  3.8633605   1.667866   3.856255  0.8473967</span></span>
<span><span class="co">#&gt; GSM1051528  3.8147481   1.599476   3.748134  0.8744509</span></span>
<span><span class="co">#&gt; GSM1051529  0.9114222   1.886485   1.997508  0.8602605</span></span>
<span><span class="co">#&gt; GSM1051530  4.3556358   2.324293   4.553996  0.8814345</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="step-2-1">Step 2<a class="anchor" aria-label="anchor" href="#step-2-1"></a>
</h3>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">hdmax2_step2</span> <span class="op">=</span> <span class="fu">hdmax2</span><span class="fu">::</span><span class="fu"><a href="../reference/estimate_effect.html">estimate_effect</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">hdmax2_step1</span>,</span>
<span>                                       m <span class="op">=</span> <span class="va">mediators_top10</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="vizualisation-of-results-1">Vizualisation of results<a class="anchor" aria-label="anchor" href="#vizualisation-of-results-1"></a>
</h3>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org" class="external-link">ggplot2</a></span><span class="op">)</span></span>
<span><span class="fu">hdmax2</span><span class="fu">::</span><span class="fu"><a href="../reference/plot_hdmax2.html">plot_hdmax2</a></span><span class="op">(</span><span class="va">hdmax2_step2</span>, plot_type <span class="op">=</span> <span class="st">"all_plot"</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "hdmax2 plot for univariate exposome"</span></span></code></pre></div>
<p><img src="hdmax2_tutorial_files/figure-html/plot%20results%20univar%20-1.png" width="960"></p>
<pre><code><span><span class="co">#&gt; TableGrob (2 x 2) "arrange": 4 grobs</span></span>
<span><span class="co">#&gt;   z     cells    name           grob</span></span>
<span><span class="co">#&gt; 1 1 (1-1,1-1) arrange gtable[layout]</span></span>
<span><span class="co">#&gt; 2 2 (1-1,2-2) arrange gtable[layout]</span></span>
<span><span class="co">#&gt; 3 3 (2-2,1-1) arrange gtable[layout]</span></span>
<span><span class="co">#&gt; 4 4 (2-2,2-2) arrange gtable[layout]</span></span></code></pre>
<ul>
<li><p><strong>A</strong> Estimates of indirect effect (ACME) and
<strong>B</strong> proportions of mediated effect (PM) for the top 10
mediators. The effect estimate is represented by a dot and its 95% CI by
the bar. Symbols correspond to the significance cut off of 5% (square
for p-value <span class="math inline">\(\geq 0.05\)</span>, circle
p-value <span class="math inline">\(&lt; 0.05\)</span>). Colors
correspond to the sign of the effect (green for estimated effect <span class="math inline">\(\leq 0\)</span> , red for estimated effect <span class="math inline">\(&gt; 0\)</span>).</p></li>
<li><p><strong>C</strong> Effect sizes of Overall Direct Effect (ODE),
Overall Indirect Effect (OIE) and Overall Total Effect (OTE). Error bars
correspond to standard deviation (ODE and OTE) or confidence interval
(OIE).</p></li>
<li><p><strong>D</strong> Indirect effect sizes for the selected
mediators. Black corresponds to the ACME, violet to the effect of
exposure <span class="math inline">\(X\)</span> on mediator
<strong>M</strong>, and blue corresponds to the effect of mediator
<strong>M</strong> on outcome <span class="math inline">\(Y\)</span>.</p></li>
</ul>
<p>In the <code>plot_hdmax2</code> function it is possible to produce
the 4-plots set or each individual plot with <em>plot_type</em>
argument.</p>
</div>
</div>
<div class="section level2">
<h2 id="second-example-multivariate-exposure-and-continuous-outcome">Second Example: <em>multivariate exposure</em> and <em>continuous
outcome</em><a class="anchor" aria-label="anchor" href="#second-example-multivariate-exposure-and-continuous-outcome"></a>
</h2>
<p>For this example we use a <strong>two variables exposure</strong>
(continuous and binary), a <strong>continuous outcome</strong> and
<strong>two adjustment factors</strong> (age and gender).</p>
<div class="section level3">
<h3 id="step-1-1">Step 1<a class="anchor" aria-label="anchor" href="#step-1-1"></a>
</h3>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## multivariate Exposures (continuous + binary)</span></span>
<span><span class="va">X</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span>X1 <span class="op">=</span> <span class="va">simu_data</span><span class="op">$</span><span class="va">X_continuous</span>, X2 <span class="op">=</span> <span class="va">simu_data</span><span class="op">$</span><span class="va">X_binary</span><span class="op">)</span></span>
<span><span class="va">covar</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/cbind.html" class="external-link">cbind</a></span><span class="op">(</span><span class="va">simu_data</span><span class="op">$</span><span class="va">age</span>, <span class="va">simu_data</span><span class="op">$</span><span class="va">gender</span><span class="op">)</span></span>
<span><span class="va">covar</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html" class="external-link">as.data.frame</a></span><span class="op">(</span><span class="va">covar</span><span class="op">)</span></span>
<span></span>
<span><span class="va">hdmax2_step1</span> <span class="op">=</span> <span class="fu">hdmax2</span><span class="fu">::</span><span class="fu"><a href="../reference/run_AS.html">run_AS</a></span><span class="op">(</span>exposure <span class="op">=</span> <span class="va">X</span> ,</span>
<span>                              outcome <span class="op">=</span> <span class="va">simu_data</span><span class="op">$</span><span class="va">Y_continuous</span>,</span>
<span>                              M <span class="op">=</span> <span class="va">simu_data</span><span class="op">$</span><span class="va">M2</span>,</span>
<span>                              K <span class="op">=</span> <span class="va">K</span>,</span>
<span>                              covar <span class="op">=</span> <span class="va">covar</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">hdmax2_step1</span><span class="op">$</span><span class="va">max2_pvalues</span><span class="op">)</span></span>
<span><span class="co">#&gt;    cg00005740    cg00006787    cg00007032    cg00008612    cg00009871 </span></span>
<span><span class="co">#&gt;  3.847528e-20  1.864280e-14 3.894362e-155  1.964262e-12  4.594620e-10 </span></span>
<span><span class="co">#&gt;    cg00009922 </span></span>
<span><span class="co">#&gt;  1.071120e-07</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="selection-of-a-subset-of-mediators-2">Selection of a subset of mediators<a class="anchor" aria-label="anchor" href="#selection-of-a-subset-of-mediators-2"></a>
</h3>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Selecting top 10 mediators</span></span>
<span><span class="va">mediators_top10</span> <span class="op">=</span> <span class="va">simu_data</span><span class="op">$</span><span class="va">M2</span><span class="op">[</span>,<span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/sort.html" class="external-link">sort</a></span><span class="op">(</span><span class="va">hdmax2_step1</span><span class="op">$</span><span class="va">max2_pvalues</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">10</span><span class="op">]</span><span class="op">)</span><span class="op">]</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">mediators_top10</span><span class="op">)</span></span>
<span><span class="co">#&gt;            cg00007032 cg00022633 cg00025981 cg00031759 cg00028749 cg00019093</span></span>
<span><span class="co">#&gt; GSM1051525   1.764355   3.042535   1.987983   2.229753   2.020358   2.021450</span></span>
<span><span class="co">#&gt; GSM1051526   6.324724   7.291695   5.055701   5.322832   5.374205   4.562818</span></span>
<span><span class="co">#&gt; GSM1051527   7.849346   8.603409   6.022010   6.260643   6.420444   5.316097</span></span>
<span><span class="co">#&gt; GSM1051528   6.590844   7.295384   5.110906   5.310325   5.434230   4.522164</span></span>
<span><span class="co">#&gt; GSM1051529   2.352573   3.809111   2.515193   2.784983   2.567937   2.521325</span></span>
<span><span class="co">#&gt; GSM1051530   7.989509   9.701877   6.659326   7.033794   7.011228   6.110346</span></span>
<span><span class="co">#&gt;            cg00014667 cg00012692 cg00018896 cg00005740</span></span>
<span><span class="co">#&gt; GSM1051525   2.104173   1.284970   2.200426   2.052600</span></span>
<span><span class="co">#&gt; GSM1051526   5.692560   4.653031   5.962785   5.143245</span></span>
<span><span class="co">#&gt; GSM1051527   6.884147   5.781892   7.210383   6.173062</span></span>
<span><span class="co">#&gt; GSM1051528   5.896802   4.873684   6.173285   5.317930</span></span>
<span><span class="co">#&gt; GSM1051529   2.569591   1.724780   2.615735   2.445169</span></span>
<span><span class="co">#&gt; GSM1051530   7.019162   5.875805   7.400408   6.219636</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="step-2-2">Step 2<a class="anchor" aria-label="anchor" href="#step-2-2"></a>
</h3>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">hdmax2_step2</span> <span class="op">=</span> <span class="fu">hdmax2</span><span class="fu">::</span><span class="fu"><a href="../reference/estimate_effect.html">estimate_effect</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">hdmax2_step1</span>,</span>
<span>                                       m <span class="op">=</span> <span class="va">mediators_top10</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="vizualisation-of-results-2">Vizualisation of results<a class="anchor" aria-label="anchor" href="#vizualisation-of-results-2"></a>
</h3>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org" class="external-link">ggplot2</a></span><span class="op">)</span></span>
<span><span class="fu">hdmax2</span><span class="fu">::</span><span class="fu"><a href="../reference/plot_hdmax2.html">plot_hdmax2</a></span><span class="op">(</span><span class="va">hdmax2_step2</span>, plot_type<span class="op">=</span> <span class="st">"all_plot"</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "hdmax2 plot for multivariate exposome"</span></span></code></pre></div>
<p><img src="hdmax2_tutorial_files/figure-html/plot%20results%20multivar%20-1.png" width="960"><img src="hdmax2_tutorial_files/figure-html/plot%20results%20multivar%20-2.png" width="960"></p>
<pre><code><span><span class="co">#&gt; $X1</span></span>
<span><span class="co">#&gt; TableGrob (2 x 2) "arrange": 4 grobs</span></span>
<span><span class="co">#&gt;   z     cells    name           grob</span></span>
<span><span class="co">#&gt; 1 1 (1-1,1-1) arrange gtable[layout]</span></span>
<span><span class="co">#&gt; 2 2 (1-1,2-2) arrange gtable[layout]</span></span>
<span><span class="co">#&gt; 3 3 (2-2,1-1) arrange gtable[layout]</span></span>
<span><span class="co">#&gt; 4 4 (2-2,2-2) arrange gtable[layout]</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $X2</span></span>
<span><span class="co">#&gt; TableGrob (2 x 2) "arrange": 4 grobs</span></span>
<span><span class="co">#&gt;   z     cells    name           grob</span></span>
<span><span class="co">#&gt; 1 1 (1-1,1-1) arrange gtable[layout]</span></span>
<span><span class="co">#&gt; 2 2 (1-1,2-2) arrange gtable[layout]</span></span>
<span><span class="co">#&gt; 3 3 (2-2,1-1) arrange gtable[layout]</span></span>
<span><span class="co">#&gt; 4 4 (2-2,2-2) arrange gtable[layout]</span></span></code></pre>
<p>Please refer to previous example for figures comments.</p>
</div>
</div>
<div class="section level2">
<h2 id="helper-functions">Helper Functions<a class="anchor" aria-label="anchor" href="#helper-functions"></a>
</h2>
<p>In this vignette, we also provide a series of helper function to
process the data.</p>
<div class="section level3">
<h3 id="how-to-analyse-agregated-methylated-regions">How to analyse agregated methylated regions ?<a class="anchor" aria-label="anchor" href="#how-to-analyse-agregated-methylated-regions"></a>
</h3>
<p>We simulated 500 samples and 1000 potential DNA methylation
mediators, with various a binary exposure (smoking status of mothers)
and continuous outcomes (birth weight).</p>
<div class="section level4">
<h4 id="identifying-aggregated-mediator-regions-amr">Identifying aggregated mediator regions (AMR)<a class="anchor" aria-label="anchor" href="#identifying-aggregated-mediator-regions-amr"></a>
</h4>
<p>Identify Aggregated Methylated regions (AMR) with
<code>AMR_search</code> function which uses from the <span class="math inline">\(P\)</span>-values from <em>max-squared</em> test
compute in <code>run_AS</code> function and using a adapted
<em>comb-p</em> method (<em>comb-p</em> is a tool that manipulates BED
files of possibly irregularly spaced <span class="math inline">\(P\)</span>-values and calculates auto-correlation,
combines adjacent <span class="math inline">\(P\)</span>-values,
performs false discovery adjustment, finds regions of enrichment and
assigns significance to those regions). AMR identification could be
useful in Epigenomic Wide Association Studies (EWAS) when single CpG
mediation is unsuccessful, or could be complementary analysis.</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">helper_ex</span> <span class="op">=</span> <span class="fu">hdmax2</span><span class="fu">::</span><span class="va"><a href="../reference/helper_ex.html">helper_ex</a></span></span>
<span></span>
<span><span class="co">#Artificial reduction of dataset size to pass the github action check when building hdmax2 website</span></span>
<span><span class="va">helper_ex</span><span class="op">$</span><span class="va">methylation</span> <span class="op">=</span> <span class="va">helper_ex</span><span class="op">$</span><span class="va">methylation</span><span class="op">[</span> , <span class="fl">800</span><span class="op">:</span><span class="fl">1000</span><span class="op">]</span></span>
<span><span class="va">helper_ex</span><span class="op">$</span><span class="va">annotation</span> <span class="op">=</span> <span class="va">helper_ex</span><span class="op">$</span><span class="va">annotation</span><span class="op">[</span><span class="fl">800</span><span class="op">:</span><span class="fl">1000</span>, <span class="op">]</span></span>
<span></span>
<span></span>
<span><span class="va">pc</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/prcomp.html" class="external-link">prcomp</a></span><span class="op">(</span><span class="va">helper_ex</span><span class="op">$</span><span class="va">methylation</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="op">(</span><span class="va">pc</span><span class="op">$</span><span class="va">sdev</span><span class="op">^</span><span class="fl">2</span><span class="op">/</span><span class="fu"><a href="https://rdrr.io/r/base/sum.html" class="external-link">sum</a></span><span class="op">(</span><span class="va">pc</span><span class="op">$</span><span class="va">sdev</span><span class="op">^</span><span class="fl">2</span><span class="op">)</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">15</span><span class="op">]</span>,</span>
<span>     xlab <span class="op">=</span> <span class="st">'Principal Component'</span>,</span>
<span>     ylab <span class="op">=</span> <span class="st">"Explained variance"</span>,</span>
<span>     col <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">3</span><span class="op">)</span>, <span class="fl">2</span>, <span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">16</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="hdmax2_tutorial_files/figure-html/AMR%20-1.png" width="700"></p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># chosen number of dimension</span></span>
<span><span class="va">K</span><span class="op">=</span><span class="fl">5</span></span>
<span></span>
<span><span class="co">## run hdmax2 step1</span></span>
<span><span class="va">hdmax2_step1</span> <span class="op">=</span> <span class="fu">hdmax2</span><span class="fu">::</span><span class="fu"><a href="../reference/run_AS.html">run_AS</a></span><span class="op">(</span>exposure <span class="op">=</span> <span class="va">helper_ex</span><span class="op">$</span><span class="va">exposure</span>,</span>
<span>                              outcome <span class="op">=</span> <span class="va">helper_ex</span><span class="op">$</span><span class="va">phenotype</span>,</span>
<span>                              M <span class="op">=</span> <span class="va">helper_ex</span><span class="op">$</span><span class="va">methylation</span>, </span>
<span>                              K <span class="op">=</span> <span class="va">K</span><span class="op">)</span></span>
<span></span>
<span><span class="co">##Detecting AMR</span></span>
<span><span class="va">seed</span> <span class="op">=</span> <span class="fl">0.6</span> <span class="co">#Careful to change this parameter when working with real data</span></span>
<span><span class="va">res.amr_search</span> <span class="op">=</span> <span class="fu">hdmax2</span><span class="fu">::</span><span class="fu"><a href="../reference/AMR_search.html">AMR_search</a></span><span class="op">(</span>chr <span class="op">=</span> <span class="va">helper_ex</span><span class="op">$</span><span class="va">annotation</span><span class="op">$</span><span class="va">chr</span>,</span>
<span>                                    start <span class="op">=</span> <span class="va">helper_ex</span><span class="op">$</span><span class="va">annotation</span><span class="op">$</span><span class="va">start</span>,</span>
<span>                                    end <span class="op">=</span> <span class="va">helper_ex</span><span class="op">$</span><span class="va">annotation</span><span class="op">$</span><span class="va">end</span>,</span>
<span>                                    pval <span class="op">=</span> <span class="va">hdmax2_step1</span><span class="op">$</span><span class="va">max2_pvalues</span>,</span>
<span>                                    cpg <span class="op">=</span> <span class="va">helper_ex</span><span class="op">$</span><span class="va">annotation</span><span class="op">$</span><span class="va">cpg</span>,</span>
<span>                                    seed <span class="op">=</span> <span class="va">seed</span>, </span>
<span>                                    nCores <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span>
<span></span>
<span><span class="va">res.amr_search</span><span class="op">$</span><span class="va">res</span></span>
<span><span class="co">#&gt;   chr   start     end           p        fdr</span></span>
<span><span class="co">#&gt; 2   1 1326166 1326221 0.005626678 0.01688003</span></span>
<span><span class="co">#&gt; 1   1 1310734 1310988 0.019799860 0.02138751</span></span>
<span><span class="co">#&gt; 3   1 1406862 1407016 0.021387509 0.02138751</span></span>
<span></span>
<span><span class="va">res.arm_build</span> <span class="op">=</span> <span class="fu">hdmax2</span><span class="fu">::</span><span class="fu"><a href="../reference/AMR_build.html">AMR_build</a></span><span class="op">(</span><span class="va">res.amr_search</span>, </span>
<span>                                  methylation <span class="op">=</span> <span class="va">helper_ex</span><span class="op">$</span><span class="va">methylation</span>, </span>
<span>                                  nb_cpg <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span>
<span></span>
<span><span class="co">#List of DMR selected</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">res.arm_build</span><span class="op">$</span><span class="va">res</span><span class="op">)</span></span>
<span><span class="co">#&gt;    DMR chr   start     end          p        fdr nb</span></span>
<span><span class="co">#&gt; 1 DMR1   1 1310734 1310988 0.01979986 0.02138751  2</span></span>
<span><span class="co">#&gt; 2 DMR2   1 1406862 1407016 0.02138751 0.02138751  2</span></span>
<span></span>
<span><span class="co">##CpG in the DMR</span></span>
<span><span class="va">res.arm_build</span><span class="op">$</span><span class="va">CpG_for_each_AMR</span></span>
<span><span class="co">#&gt; $DMR1</span></span>
<span><span class="co">#&gt; [1] "cg11518257" "cg15617543"</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $DMR2</span></span>
<span><span class="co">#&gt; [1] "cg10228629" "cg06377929"</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="quantifying-indirect-effects">Quantifying indirect effects<a class="anchor" aria-label="anchor" href="#quantifying-indirect-effects"></a>
</h4>
<p>Like with single mediators analysis <em>esimate_effect</em> function
could be use to estimate different effects of AMR. Also
<code>plot_hdmax2</code> can be applied to step2 results.</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## run hdmax2 step2</span></span>
<span><span class="va">object</span> <span class="op">=</span> <span class="va">hdmax2_step1</span></span>
<span><span class="va">mediators_top10</span> <span class="op">=</span> <span class="va">helper_ex</span><span class="op">$</span><span class="va">methylation</span><span class="op">[</span>,<span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/sort.html" class="external-link">sort</a></span><span class="op">(</span><span class="va">hdmax2_step1</span><span class="op">$</span><span class="va">max2_pvalues</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">10</span><span class="op">]</span><span class="op">)</span><span class="op">]</span></span>
<span><span class="va">m</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">as.matrix</a></span><span class="op">(</span><span class="va">res.arm_build</span><span class="op">$</span><span class="va">AMR_mean</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## selected mediators effects estimation</span></span>
<span></span>
<span><span class="va">hdmax2_step2</span> <span class="op">=</span> <span class="fu">hdmax2</span><span class="fu">::</span><span class="fu"><a href="../reference/estimate_effect.html">estimate_effect</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">hdmax2_step1</span>,</span>
<span>                                       m <span class="op">=</span> <span class="va">m</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org" class="external-link">ggplot2</a></span><span class="op">)</span></span>
<span><span class="fu">hdmax2</span><span class="fu">::</span><span class="fu"><a href="../reference/plot_hdmax2.html">plot_hdmax2</a></span><span class="op">(</span><span class="va">hdmax2_step2</span>, plot_type<span class="op">=</span> <span class="st">"all_plot"</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "hdmax2 plot for univariate exposome"</span></span></code></pre></div>
<p><img src="hdmax2_tutorial_files/figure-html/step2-1.png" width="960"></p>
<pre><code><span><span class="co">#&gt; TableGrob (2 x 2) "arrange": 4 grobs</span></span>
<span><span class="co">#&gt;   z     cells    name           grob</span></span>
<span><span class="co">#&gt; 1 1 (1-1,1-1) arrange gtable[layout]</span></span>
<span><span class="co">#&gt; 2 2 (1-1,2-2) arrange gtable[layout]</span></span>
<span><span class="co">#&gt; 3 3 (2-2,1-1) arrange gtable[layout]</span></span>
<span><span class="co">#&gt; 4 4 (2-2,2-2) arrange gtable[layout]</span></span></code></pre>
</div>
</div>
<div class="section level3">
<h3 id="how-to-select-potential-mediators-using-q-values-and-fdr-control">How to select potential mediators using q-values and FDR control
?<a class="anchor" aria-label="anchor" href="#how-to-select-potential-mediators-using-q-values-and-fdr-control"></a>
</h3>
<p>Several methods can be use to select potential mediators from step1
results of our method, in our main use cases we simply select top 10
mediators to simplify the narrative. Among available methods to select
mediators from mediation test <span class="math inline">\(P\)</span>-values, we can use FDR (False Discovery
Rate) control. <span class="math inline">\(Q\)</span>-value is obtained
from <em>max-squared</em> test <span class="math inline">\(P\)</span>-values from step 1 with
<code>fdrtools::fdrtools</code> function then bounded by chosen
threshold.</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## run hdmax2 step1</span></span>
<span><span class="va">hdmax2_step1</span> <span class="op">=</span> <span class="fu">hdmax2</span><span class="fu">::</span><span class="fu"><a href="../reference/run_AS.html">run_AS</a></span><span class="op">(</span>exposure <span class="op">=</span> <span class="va">simu_data</span><span class="op">$</span><span class="va">X_binary</span>,</span>
<span>                 outcome <span class="op">=</span> <span class="va">simu_data</span><span class="op">$</span><span class="va">Y_continuous</span>,</span>
<span>                 M <span class="op">=</span> <span class="va">simu_data</span><span class="op">$</span><span class="va">M1</span>, </span>
<span>                 K <span class="op">=</span> <span class="va">K</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## Select candidate mediator  </span></span>
<span><span class="va">q_values</span> <span class="op">=</span> <span class="fu">hdmax2</span><span class="fu">::</span><span class="fu"><a href="../reference/hdmax2_qvalue.html">hdmax2_qvalue</a></span><span class="op">(</span><span class="va">hdmax2_step1</span><span class="op">$</span><span class="va">max2_pvalues</span><span class="op">)</span></span>
<span><span class="va">candidate_mediator</span> <span class="op">&lt;-</span> <span class="va">q_values</span><span class="op">$</span><span class="va">qv</span><span class="op">[</span><span class="va">q_values</span><span class="op">$</span><span class="va">qv</span><span class="op">&lt;=</span> <span class="fl">0.2</span><span class="op">]</span> <span class="co"># In this example, we will consider FDR levels &lt;20%.</span></span>
<span><span class="va">candidate_mediator</span></span>
<span><span class="co">#&gt;   cg00018896   cg00019093   cg00022633   cg00024247   cg00025981   cg00028749 </span></span>
<span><span class="co">#&gt; 1.097593e-02 3.839397e-10 3.379984e-15 2.861555e-02 3.977704e-06 2.864961e-02 </span></span>
<span><span class="co">#&gt;   cg00031759   cg00035636   cg00037930   cg00047079   cg00049102   cg00049382 </span></span>
<span><span class="co">#&gt; 1.067235e-04 2.138923e-08 9.146395e-02 2.572296e-02 8.596995e-09 2.682088e-04 </span></span>
<span><span class="co">#&gt;   cg00049616   cg00241432   cg00345862   cg00465739   cg01005425 </span></span>
<span><span class="co">#&gt; 2.397814e-05 1.455775e-01 7.647113e-02 1.761855e-01 1.300586e-01</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="how-to-transform-a-categorial-variable-into-an-ordered-factor">How to transform a categorial variable into an ordered factor?<a class="anchor" aria-label="anchor" href="#how-to-transform-a-categorial-variable-into-an-ordered-factor"></a>
</h3>
<p>When categorical variables are used as exposure variables,
<code>hdmax2</code> uses the first category as a reference (intercept)
to calculate the effects associated with the variable’s other
categories. The functions <code>HDMAX2::run_AS</code> and
<code>HDMAX2::estimate_effect</code> will transform the character vector
you have used (if any) into a factor, with an arbitrary ordering of the
categories. If order is important to you, here’s a simple way to turn
your character vector into a factor and order the categories as you
wish. You can then use this variable as input to the <code>hdmax2</code>
functions.</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># inverse alphabetical order for example</span></span>
<span><span class="va">X</span> <span class="op">=</span> <span class="va">simu_data</span><span class="op">$</span><span class="va">X_categorial</span></span>
<span><span class="va">X</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">factor</a></span><span class="op">(</span><span class="va">X</span>, ordered <span class="op">=</span> <span class="cn">TRUE</span>, levels <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"D"</span>,<span class="st">"C"</span>,<span class="st">"B"</span>,<span class="st">"A"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="how-to-handle-adding-an-additional-set-of-covariates-to-the-second-association-study">How to handle adding an additional set of covariates to the second
association study?<a class="anchor" aria-label="anchor" href="#how-to-handle-adding-an-additional-set-of-covariates-to-the-second-association-study"></a>
</h3>
<p>It is possible to add a second adjustment factors set in association
study between potential mediators and outcome if it makes sense from a
biological standpoint. Nevertheless, this additional set of adjustment
factors for the second association study must include the adjustment
factors from the first association study as design in
<code>run_AS</code> function.</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># First adjustment factors set</span></span>
<span><span class="va">covar</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html" class="external-link">as.data.frame</a></span><span class="op">(</span><span class="va">helper_ex</span><span class="op">$</span><span class="va">covariables</span><span class="op">[</span>,<span class="fl">1</span><span class="op">:</span><span class="fl">2</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="co"># Second adjustment factors set</span></span>
<span><span class="va">suppl_covar</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html" class="external-link">as.data.frame</a></span><span class="op">(</span><span class="va">helper_ex</span><span class="op">$</span><span class="va">covariables</span><span class="op">[</span>,<span class="fl">3</span><span class="op">:</span><span class="fl">4</span><span class="op">]</span><span class="op">)</span></span>
<span></span>
<span><span class="va">hdmax2_step1</span> <span class="op">=</span> <span class="fu">hdmax2</span><span class="fu">::</span><span class="fu"><a href="../reference/run_AS.html">run_AS</a></span><span class="op">(</span>exposure <span class="op">=</span> <span class="va">helper_ex</span><span class="op">$</span><span class="va">exposure</span>,</span>
<span>                              outcome <span class="op">=</span> <span class="va">helper_ex</span><span class="op">$</span><span class="va">phenotype</span>,</span>
<span>                              M <span class="op">=</span> <span class="va">helper_ex</span><span class="op">$</span><span class="va">methylation</span>,</span>
<span>                              K <span class="op">=</span> <span class="va">K</span>,</span>
<span>                              covar <span class="op">=</span> <span class="va">covar</span>,</span>
<span>                              suppl_covar <span class="op">=</span> <span class="va">suppl_covar</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
</div>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Florence Pittion, Magali Richard, Olivier Francois, Basile Jumentier.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.9.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
